listenbrainz/webserver/static/js/src/BrainzPlayer.tsx

```diff
@@ -93,6 +92,31 @@ type BrainzPlayerState = {
   continuousPlaybackTime: number;
 };
 
+/**
+ * Due to some issue with TypeScript when accessing static methods of an instance when you don't know
+ * which class it is, we have to manually determine the class of the instance and call MyClass.staticMethod().
+ * Neither instance.constructor.staticMethod() nor instance.prototype.constructor.staticMethod() work without issues.
+ * See https://github.com/Microsoft/TypeScript/issues/3841#issuecomment-337560146
+ */
+function isListenFromDatasource(
+  listen: BaseListenFormat | Listen | JSPFTrack,
+  datasource: DataSourceTypes | null
+) {
+  if (!listen || !datasource) {
+    return undefined;
+  }
+  if (datasource instanceof SpotifyPlayer) {
+    return SpotifyPlayer.isListenFromThisService(listen);
+  }
+  if (datasource instanceof YoutubePlayer) {
+    return YoutubePlayer.isListenFromThisService(listen);
+  }
+  if (datasource instanceof SoundcloudPlayer) {
```

Comment:
We can directly return without explicitly checking here since we have already confirmed with the first `if` statement about the `datasource` being one of the three or not.

Reply:
That's how it was initially, but unfortunately there is an issue with typescript that prevents us from doing it that way.

I had to make `isListenFromThisService` a static method in order to be able to use it in `getAlbumArtFromListenMetadata` in utils.tsx, where I don't have access to instances of these datasource classes.

Typescript freaks out when I try `datasource.isListenFromThisService(listen)` since it is not aware of static methods on these classes without referring to a class directly (`YoutubePlayer.isListenFromThisService(…)` works but `datasource.isListenFromThisService(…)` doesn't, and same goes for `YoutubePlayer.prototype.isListenFromThisService(…)` and `YoutubePlayer.constructor.isListenFromThisService(…)`)

It is a know TS issue which I left in a comment to above this method for posterity: https://github.com/Microsoft/TypeScript/issues/3841#issuecomment-337560146